loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,467 @@
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error calling finalize for database: %s with Result Code: %i", parentDB->getDBName(), result);
+        }
+        return result;        
+    }
+};
+
+
+//---Statement--- external function initialisation
+int __stdcall Statement::stepAsyncBody(void *param)
+{
+    //get the statement
+    Statement *s = (Statement *)param;
+
+
+    //call the internal SQLite step function, then yield the thread
+    int result = sqlite3_step(s->statementHandle); 
+    loom_thread_yield();
+
+    //mutex so we can call the delegate
+    s->stepAsyncDelegateMutex = loom_mutex_create();
+    loom_mutex_lock(s->stepAsyncDelegateMutex);
+
+    //check the result now
+    if(result == SQLITE_ROW)
+    {
+        s->_OnStatementProgressDelegate.invoke();
+    }
+    else if(result == SQLITE_DONE)
+    {
+//HACK: s->stepAsyncLuaState is not working as it can be unstable... 
+//      need another way to pass the Statement through to OnStatementComplete!!!
+lualoom_pushnative<Statement>(s->stepAsyncLuaState, s);
+s->_OnStatementCompleteDelegate.incArgCount();
+s->_OnStatementCompleteDelegate.invoke();
+lua_pop(s->stepAsyncLuaState, 1);
+    }
+    else
+    {
+        if(result == SQLITE_ERROR)
+        {
+            lmLogError(gSQLiteGroup, "Error calling step for database: %s", s->parentDB->getDBName());
+        }
+//TODO: Should we have an additiona _OnStatementErrorDelegate to call here?
+    }
+
+    //we're done now!
+    loom_mutex_unlock(s->stepAsyncDelegateMutex);
+}
+
+
+
+
+
+//---Connection--- external function / variable initialisation
+NativeDelegate Connection::_OnImportCompleteDelegate;
+
+Statement *Connection::prepare(const char *query)
+{
+    Statement *s = new Statement(this);
+
+    //prepare the database with the query provided
+    int res = sqlite3_prepare_v2(dbHandle, query, -1, &s->statementHandle, NULL);
+    if(res != SQLITE_OK)
+    {
+        lmLogError(gSQLiteGroup, "Error preparing the SQLite database: %s with message: %s", getDBName(), getErrorMessage());
+    }
+    return s;
+}
+
+Connection *Connection::open(const char *database, const char *path, int flags)
+{
+    Connection *c;
+
+    //check for valid database name
+    if((database == NULL) || (database[0] == '\0'))
+    {
+        lmLogError(gSQLiteGroup, "Invalid database name specified!");
+        return NULL;
+    }
+
+    //create the connection
+    c = new Connection();
+    c->databaseName = utString(database);
+    
+    //prep the database path
+    if((path == NULL) || (path[0] == '\0'))
+    {
+        //prefix the system writable path in front of the database name
+        c->databaseFullPath = utString(cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str());
+    }
+    else
+    {
+        //assume that the user specified a valid system writeable path!!!
+        c->databaseFullPath = utString(path);
+    }
+    c->databaseFullPath += c->databaseName;
```

Comment:
So if database starts with a /, we assume it's absolute, if not, it's just
a filename and we add the system writable path?

Or just always tries to create the straight database but if it fails then
it tries with the system path?
On 12 Feb 2015 10:59 PM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24619584>:
>
> > +    //create the connection
> > +    c = new Connection();
> > +    c->databaseName = utString(database);
> > +
> > +    //prep the database path
> > +    if((path == NULL) || (path[0] == '\0'))
> > +    {
> > +        //prefix the system writable path in front of the database name
> > +        c->databaseFullPath = utString(cocos2d::CCFileUtils::sharedFileUtils()->getWriteablePath().c_str());
> > +    }
> > +    else
> > +    {
> > +        //assume that the user specified a valid system writeable path!!!
> > +        c->databaseFullPath = utString(path);
> > +    }
> > +    c->databaseFullPath += c->databaseName;
>
> Can we switch this logic off of path containing a "/"? So if it does it
> tries to load it direct first, then tries in the writeable path?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24619584>.
>